File: source/ChickenSoftware.IntroToFSharp.Solution/ChickenSoftware.ActiveDirectory.FS/ADGroup.fs
namespace ChickenSoftware.FSharpFeatures

open System;
open System.Collections.ObjectModel
open System.DirectoryServices.AccountManagement

type ADGroupInfo = {AccountName:string; Description:string; UserId:string }

type ADGroup() = 
    member this.GetList userId = 
        use context = new PrincipalContext(ContextType.Domain, "NEWCO")
        use user = UserPrincipal.FindByIdentity(context, IdentityType.SamAccountName, userId)
        try
            user.GetGroups()
            |> Seq.filter(fun g -> g :? GroupPrincipal)
            |> Seq.sortBy(fun g -> g.Description)
            |> Seq.toArray
            |> ReadOnlyCollection<_>
            |> Some
        with
        | :? NoMatchingPrincipalException -> None

File: source/ChickenSoftware.IntroToFSharp.Solution/ChickenSoftware.ActiveDirectory/ADGroup.cs
using System;

namespace ChickenSoftware.ActiveDirectory
{
    public class ADGroup
    {

        public ADGroup(string accountName, string description, string userId)
        {
            if(string.IsNullOrEmpty(accountName))
            {
                throw new ArgumentException("accountName");

            }
            if (string.IsNullOrEmpty(description))
            {
                throw new ArgumentException("description");

            }
            if (string.IsNullOrEmpty(userId))
            {
                throw new ArgumentException("userId");

            }

            this.AccountName = accountName;
            this.Description = description;
            this.UserId = userId;
        }

        public string AccountName { get; set; }
        public string Description { get; set; }
        public string UserId { get; set; }
    }
}